### PR TITLE
chore: idx-js refactor

### PR DESCRIPTION
**Pull request by nikhilvenkatraman-okta**
Resolves: OKTA-276550

- Removed actionsTransformer as its no longer used 
- Removed code from responseTransformer where we created these actions 
- Remove currentState object to be replaced with idx object 
- Set remediations object that is used in uiSchemaTransformer to create uiSchema
- Updated tests

### CODE REVIEW

File: src/v2/models/AppState.js [hunk 1]
```diff
@@ -131,6 +131,7 @@ export default Model.extend({
     }
 
     this.set(resp);
+    this.set('__previousResponse', idx.rawIdxState);
```
Comment (magizh-okta) [inline src/v2/models/AppState.js]: @nikhilvenkatraman-okta  I dont know what changes did you make but looks like we are stil settings things the old way and not clearing the old object we set. eg) AuthenticatorObject.

Comment (nikhilvenkatraman-okta) [reply src/v2/models/AppState.js]: yeah, I only included the change to use `__previousResponse `. Still testing the change to clear appState object before we set it. Should be a small change I'll include it in the next commit.

File: src/v2/ion/responseTransformer.js [hunk 2]
```diff
@@ -145,72 +48,33 @@ const convertObjectType = (resp) => {
...
-
-  return Object.assign(restData, actions, {
-    remediation,
-  });
-
-};
-
 /**
  *
  * @param {AuthnResponse} resp authn v2 response
```
[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/v2/ion/responseTransformer.js]: updated

File: src/v2/models/AppState.js [hunk 3]
```diff
@@ -23,15 +23,12 @@ export default Model.extend({
...
     user: 'object',        // optional
     currentFormName: 'string',
+    idx: 'object',
+    remediations: 'array',
+    __previousResponse: 'object'
   },
 
   derived: {
-    remediation: {
-      deps: ['currentState'],
```
Comment (jhe-okta) [inline src/v2/models/AppState.js]: Is `currentState` still being used anywhere else? If not, we should remove the definition.

File: test/unit/helpers/xhr/v2/ENROLL_PROFILE.js [hunk 4]
```diff
@@ -70,21 +70,6 @@ const data = {
         }
       ]
     },
-    "cancel": {
```
Comment (jhe-okta) [inline test/unit/helpers/xhr/v2/ENROLL_PROFILE.js]: Is this just a clean up?

Comment (nikhilvenkatraman-okta) [reply test/unit/helpers/xhr/v2/ENROLL_PROFILE.js]: ah, i'll double check this. Might be able to revert.

File: test/unit/helpers/xhr/v2/FACTOR_REQUIRED_EMAIL.js [hunk 5]
```diff
@@ -2,89 +2,93 @@ const data = {
...
...mVuY3J5cHRpb25rZXkiLCJ2ZXIiOiIxIiwib2lkIjoiMDBvczI0VHZiWHlqOVFLSm4wZzMiLCJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiZGlyIn0..-rltUuvhZgmMFFpS.Hg6iqCkAnVmTou1Nonde81sqNgA5lBtYwdkKmuTXQBm3MvWU1YvL2uj5iqJDzk7ewxQRJLtlKZ20p4-m5aft-c4TEPZG9IeM0izo7kmksltr3zrmrzM1E_pWFHVq-1B6NoEivWqE3M4rXRvq3DSM0eWV8zNtadhxm0YT2KPzzu1wfR47azpGl6cMWUtABcnx_OaKEcGvlmFuypHglVf9Moza0Fbk-ywWHmGkMPcAsqQ6XYi6ovhyClzIFjkI515oYpLKpbzP6wr0UOceRcoHfPqLDPPCP1J_RyGzTp5yD3_gmZfTqcHRqTgTG7rYDTmjFu7EC6YhtlkQzEDp0aONdNS-vAO1oiPESG6sWXIX0ZMHIIF_8dtswnYQt9-cr-pD2lW1NILivtbv6qPGqWfi80GIQSs-Fp2Am4a_Z-Cvio85rajXHJRO_5bSMYvqazAx1oFast0BeHaiTDOanF73QWdfT1sDv_TiqbqWD7H5QU3RFMHtwYTyP99BVq1kPm3lnGzejBNh_vagUeF2dmhNl8QjF4QfeibBepsiX77LCrJGdTua5HgGeGpD2dH8DzW5JpR5jR4fUVw1hY6GxM28rOhjtRHDUZAdhRqE-D88IAPxX1kaqggGFXOYGM4EwgnQdgThj593cItRe6Yx5AdsW5A2NNh08ezEtoSVXSRD-nkug-aQptK7pIfPAxy18JEW_6eKJmsDJWG-XyR5SBEi2dtqq4PG-LVNBJzPdPXSBJ17uth0wrL_F3lKzUgLNmRXzcayD0T7BhZDRoMqgJPDrDI047gfoN7sV0zhOihuNTdVEOjpSzGCNnmXTUbmlOCESgOWNq7wJc_Q09qmyKdY5kjZRuGo2JWFnGrjOsAEVMYdNxIQGdizZxlP2OMr2h_rqKklgidhkFLn-EY0ig2uCYh6I3FHF5qHajaGjfgKh7ClgMN1lLuHrizwLt5uSCk_1QIbVwDELAVAQT8FhiZovB-AYZM-AB3uFfuJBXqT26n_dNsECrKmrSnT2BBc4pHhaKruoX_XX6j0EhXXuiEQIw.0hFF_OMVB-29RgUxmN7tHw",
```
Comment (magizh-okta) [inline test/unit/helpers/xhr/v2/FACTOR_REQUIRED_EMAIL.js]: can we make the statehandle shorter for readability

File: test/unit/spec/v2/ion/responseTransformer_spec.js [hunk 6]
```diff
@@ -1,237 +1,425 @@
...
-        'stateHandle': '01OCl7uyAUC4CUqHsObI9bvFiq01cRFgbnpJQ1bz82',
-        'expiresAt': '2018-09-17T23:08:56.000Z',
-        'step': 'FACTOR_REQUIRED',
-        'intent': 'login',
-        'submit-factor': jasmine.any(Function),
-        'cancel': jasmine.any(Function),
-        'context': jasmine.any(Function),
-        'remediation': [
+  it('converts factor required email idx object', () => {
+    const idxObjectFactorRequiredEmail  = {
```
Comment (magizh-okta) [inline test/unit/spec/v2/ion/responseTransformer_spec.js]: Wish we had jest to create snapshots. Can we move these to external helper files so it is easy to follow the test flow.

File: src/v2/ion/responseTransformer.js [hunk 7]
```diff
@@ -23,108 +23,11 @@ import { _ } from 'okta';
...
-const convertRelFormToFunction = (value) => {
-  const allCreateForms = getRelObjectByName(value, ['create-form']);
-  const actions = createActions(allCreateForms);
-  const restValue = _.omit.apply(_, [value].concat(Object.keys(actions)));
-
-  return Object.assign({}, restValue, actions);
-};
 
 /**
  * Transform 'type: "object"' by converting "rel: form" to be a function.
```
Comment (magizh-okta) [inline src/v2/ion/responseTransformer.js]: please update comment. Also add a sample input and output obj . 

eg) 

 {
  proceed: fn., 
  neededToProceed: {..},
  ....
}

to 

 
  neededToProceed: {..},
  ....
}

It is worth mentioning/capturing what we are retaining here with an example.

File: src/v2/models/AppState.js [hunk 8]
```diff
@@ -45,17 +42,20 @@ export default Model.extend({
...
     showSignoutLink: {
-      deps: ['currentState'],
-      fn: function (currentState = {}) {
+      deps: ['idx'],
+      fn: function (idx = {}) {
         const invalidSignOutSteps = ['IDENTIFY', 'ENROLL', 'SUCCESS'];
         // hide signout for IDENTIFY, ENROLL & SUCCESS step
-        return _.isFunction(currentState.cancel) && !invalidSignOutSteps.includes(currentState.step);
+        return idx.actions
+          && _.isFunction(idx.actions.cancel) && !invalidSignOutSteps.includes(idx.context.step);
```
[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/v2/models/AppState.js]: yes

File: test/unit/helpers/xhr/v2/FACTOR_REQUIRED_EMAIL.js (hunk 5 above)
Comment (nikhilvenkatraman-okta) [reply test/unit/helpers/xhr/v2/FACTOR_REQUIRED_EMAIL.js]: we can, but this is how they are going to be in the new pipeline as its a JWE

File: src/v2/models/AppState.js [hunk 9]
```diff
@@ -122,15 +122,20 @@ export default Model.extend({
...
         name: 'terminal',
         value: [],
         uiSchema: [],
       };
     }
-
+    //clear appState before setting new values
+    this.clear({silent: true});
+    // set new app state properties
     this.set(resp);
```
[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/v2/models/AppState.js]: yeah, the dynamic nature of the API response makes it a bit tricky. We can discuss more.

File: src/v2/models/AppState.js [hunk 10]
```diff
@@ -45,17 +42,20 @@ export default Model.extend({
       },
     },
     currentStep: {
-      deps: ['currentState'],
-      fn: function (currentState = {}) {
-        return currentState.step && currentState.step.toLowerCase();
+      deps: ['idx'],
+      fn: function (idx = {}) {
+        if (idx && idx.context && idx.context.step) {
```
[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/v2/models/AppState.js]: Dont think we were explicitly setting it to null before. If idx.context is present we'll have access to idx.context.step. If its not set i think it'll be undefined.

File: test/unit/spec/v2/ion/uiSchemaTransformer_spec.js [hunk 11]
```diff
@@ -1,223 +1,792 @@
...
+      'context':{
+        'stateHandle':'01r2p5S9qaAjESMFuPzt7r3ZMcZZQ_vvS0Tzg56ajB',
+        'version':'1.0.0',
+        'expiresAt':'2019-07-24T21:25:33.000Z',
+        'step':'PROFILE_REQUIRED',
+        'intent':'LOGIN'
+      },
+      'rawIdxState':rawUserEnrollResponse
+    };
+    const result = _.compose(uiSchemaTransformer, responseTransformer)(transformedResponse);
```
Comment (magizh-okta) [inline test/unit/spec/v2/ion/uiSchemaTransformer_spec.js]: We seem to be testing both uiSchemaTransformer, responseTransformer here. Better to just test uiSchemaTransformer ?

**Comment by nikhilvenkatraman-okta** [comment]
created https://oktainc.atlassian.net/browse/OKTA-286273 to address name space issue in appstate

**Pull request merged**